slackapi/deno-slack-sdk
# Add support for block suggestion handlers

###  Summary

This PR adds a new method available to app functions: `addBlockSuggestionHandler`.

It is a copy-paste of the existing Block Actions handler support, basically.

After merging this, we will have in place all of the available interactivity handlers. Thus, after this gets merged would be a good opportunity to review all the interactivity code and see how we should refactor it.

## Testing

1. Pull down this branch, and point a hermes app's `import_map.json` to the location of this checked out branch + `/src/`, e.g. `file:///Users/fmaj/src/deno-slack-sdk/src/`
2. Add an external data source select menu to a message or view your app creates. Perhaps it can post a message that includes this in its Block Kit payload:

```ts
      {
        type: "input",
        block_id: "ext_select_block",
        optional: true,
        element: {
          type: "external_select",
          action_id: "ext_select_input",
          placeholder: {
            type: "plain_text",
            text: "Inspire",
          },
        },
        label: {
          type: "plain_text",
          text: "Inspirational Quote",
        },
      },
```

3. Register a block suggestion handler in your app to respond to interactions with the above drop down menu. The following example keys off of the above block kit snippet's `action_id`:

```ts
export default SlackFunction(...).addBlockSuggestionHandler(
  "ext_select_input", // <-- the action id of the select drop down menu
  async (args) => { // <-- `args` here should be nicely typed, and block_suggestion-specific payload bits available under `args.body`
    console.log('BLOCK SUGGESTION HANDLER, ', JSON.stringify(args.body, null, 2));
    // Fetch an inspirational quote
    const apiResp = await fetch("https://motivational-quote-api.herokuapp.com/quotes");
    const quotes = await apiResp.json();
    console.log('Returning', quotes.length, 'quotes');
    const opts = {
      "options": quotes.map((q) => ({value: `${q.id}`, text: {type:"plain_text", text: q.quote.slice(0,70)}})) // <-- watch out! options have a max text length of 75 characters
    };
    return opts;
  }
);
```
